services/miner/api.go

```diff
@@ -37,32 +38,149 @@ func (m *Miner) APIs() []api.API {
 	}
 }
 
+type MiningStats struct {
+	LastGBTTime          time.Time `json:"last_gbt_time"`
+	LastSubmit           time.Time `json:"last_submit_time"`
+	Last100Gbts          []int64   `json:"-"`
+	Last100GbtAvgTime    float64   `json:"last_100_gbt_avg_time"`
+	Last100Submits       []int64   `json:"-"`
+	Last100SubmitAvgTime float64   `json:"last_100_submit_avg_time"`
+	SubmitAvgTime        float64   `json:"submit_avg_time"`
+	GbtAvgTime           float64   `json:"gbt_avg_time"`
+	MaxGbtTime           float64   `json:"max_gbt_time"`
+	MaxGbtTimeLongpollid string    `json:"max_gbt_time_longpollid"`
+	MaxSubmitTime        float64   `json:"max_submit_time"`
+	MaxSubmitTimeHash    string    `json:"max_submit_time_hash"`
+	TotalGbts            int64     `json:"total_gbts"`
+	TotalGbtRequests     int64     `json:"total_gbt_requests"`
+	TotalEmptyGbts       int64     `json:"total_empty_gbts"`
+	TotalSubmits         int64     `json:"total_submits"`
+	LastTxEmptyTime      int64     `json:"-"`
+	TxEmptyAvgTime       float64   `json:"tx_empty_avg_time"`
+	TxEmptyErrs          float64   `json:"tx_empty_errs"`
+}
+
 type PublicMinerAPI struct {
-	miner      *Miner
-	lastSubmit time.Time
+	miner *Miner
+	stats MiningStats
 }
 
 func NewPublicMinerAPI(m *Miner) *PublicMinerAPI {
-	pmAPI := &PublicMinerAPI{miner: m, lastSubmit: time.Now()}
+	pmAPI := &PublicMinerAPI{miner: m, stats: MiningStats{
+		LastSubmit:     time.Now(),
+		LastGBTTime:    time.Now(),
+		Last100Gbts:    make([]int64, 0),
+		Last100Submits: make([]int64, 0),
+	}}
 	return pmAPI
 }
 
-//func (api *PublicMinerAPI) GetBlockTemplate(request *mining.TemplateRequest) (interface{}, error){
+func (api *PublicMinerAPI) StatsGbtTxEmptyErr() {
+	api.stats.TxEmptyErrs++
```

Comment:
the naming is not properly used, `NonTx` or `EmptyGbt` and It's not an error. it a status of the situation,  for example, using names like `statsEmptyGbt()` or` statsGbtNonTx()` maybe better.